R/bootstrap-model.R

```diff
@@ -0,0 +1,668 @@
+  checkmate::assert_number(seed)
+
+  boot_dir <- get_output_dir(.boot_run, .check_exists = FALSE)
+  boot_data_dir <- file.path(boot_dir, "data")
+
+  # Bootstrap directory setup
+  if(!fs::dir_exists(boot_dir) || .overwrite == TRUE){
+    if(fs::dir_exists(boot_dir)) fs::dir_delete(boot_dir)
+    fs::dir_create(boot_dir)
+    fs::dir_create(boot_data_dir)
+    # Store data within run folder and gitignore
+    ##### TODO: UPDATE TO IGNORE MODEL FILES...
+    writeLines("/data\n*.ctl\n*.mod\n*.yaml", file.path(boot_dir, ".gitignore"))
+
+    # New model setup
+    n_seq <- seq(n)
+    mod_names <- purrr::map_chr(n_seq, max_char = nchar(n), pad_left)
+    mod_paths <- file.path(boot_dir, mod_names)
+    orig_mod <- read_model(get_based_on(.boot_run))
+
+    # TODO: only include subjects that entered the original problem
+    starting_data <- nm_data(orig_mod) %>% suppressMessages()
+    if(!is.null(strat_cols)){
+      checkmate::assert_true(strat_cols %in% names(starting_data))
+    }
+
+    boot_args <- list(
+      boot_run = .boot_run,
+      all_mod_names = mod_names,
+      orig_mod_path = get_model_path(orig_mod),
+      orig_mod_id = get_model_id(orig_mod),
+      orig_data = starting_data,
+      strat_cols = strat_cols,
+      seed = seed,
+      n_samples = n,
+      boot_dir = boot_dir,
+      boot_data_dir = boot_data_dir,
+      overwrite = .overwrite
+    )
+
+    # Create model object per boot run
+    boot_models <- purrr::map(mod_paths, make_boot_run, boot_args)
+    make_boot_spec(boot_models, boot_args)
+
+    # Garbage collect - may help after handling many (potentially large) datasets
+    #  - It can be useful to call gc() after a large object has been removed, as
+    #    this may prompt R to return memory to the operating system.
+    gc()
+  }else{
+    rlang::abort(
+      c(
+        glue("Bootstrap run has already been set up at `{boot_dir}`"),
+        "pass `.overwrite = TRUE` to overwrite"
+      )
+    )
+  }
+  return(invisible(.boot_run))
+}
+
+
+#' Set up a single bootstrap model run
+#'
+#' @param mod_path absolute model path (no file extension) of a bootstrap model run.
+#' @param boot_args list of parameters needed to create a bootstrap model run.
+#'
+#' @keywords internal
+make_boot_run <- function(mod_path, boot_args){
+
+  # Update the user every 100 runs (plus beginning and end)
+  new_mod_index <- which(basename(mod_path) == boot_args$all_mod_names)
+  if(new_mod_index == 1 || new_mod_index == length(boot_args$all_mod_names) ||
+     new_mod_index %% 100 == 0){
+    verbose_msg(glue("Sampling run {new_mod_index}/{length(boot_args$all_mod_names)}"))
+  }
+
+  # Copy over control stream
+  #  - Cant use copy_model_from, as we want these individual model runs to be
+  #    regular `bbi_base_model` objects
+  orig_mod_path <- boot_args$orig_mod_path
+  mod_path_ext <- paste0(mod_path, ".", fs::path_ext(orig_mod_path))
+  fs::file_copy(
+    orig_mod_path, mod_path_ext, overwrite = boot_args$overwrite
```

Comment:
`orig_mod_path` points to the original model (e.g., 106.ctl or 102.ctl). This should point to the base boot model (106-boot.ctl or 102-boot.ctl`). Only the `-boot` versions will have `$TABLE` and `$COV` removed if requested.